### PR TITLE
Makefile for host and other changes

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
Hi Jeremy,

I hope you're fine.
If you have some spare time, can you please have a look on these changes?

Thanks
--
Léo

File: host/image.h [hunk 1]
```diff
@@ -32,13 +32,13 @@
...
 
-#include <setjmp.h>
+#include <efisetjmp.h>
 #include <ewdrv.h>
 
 typedef struct image {
 	EFI_LOADED_IMAGE prot;
 	void *data;
-	jmp_buf jmp;
+	jmp_buf jmp[1];
```
Comment (jeremy-compostella) [inline host/image.h:41]: Why ?

File: host/main.c [hunk 2]
```diff
@@ -67,7 +67,7 @@ static ewdrv_t *host_drivers[] = {
 ewdrv_t **ew_drivers = host_drivers;
 
 typedef EFI_STATUS (*efi_main_t)(EFI_HANDLE image, EFI_SYSTEM_TABLE *st);
-static jmp_buf jmp;
+static jmp_buf jmp[1];
```
Comment (jeremy-compostella) [inline host/main.c:74]: Why ?

File: host/terminal_conin.c [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			    VOID *Context)
+{
+	int n;
+	char buf[3];
+	EFI_SYSTEM_TABLE *st = (EFI_SYSTEM_TABLE *)Context;
+
+	n = read(STDIN_FILENO, buf, sizeof(buf)*sizeof(buf[0]));
+	if (n > 0)
+		uefi_call_wrapper(st->BootServices->SignalEvent, 1, st->ConIn->WaitForKey);
+
```
Comment (jeremy-compostella) [inline host/terminal_conin.c]: remove extra empty line

File: host/terminal_conin.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2016, Intel Corporation
```
Comment (jeremy-compostella) [inline host/terminal_conin.h]: 2017

File: host/terminal_conin.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	if (old_term == NULL)
+		return EFI_LOAD_ERROR;
+
+	ret = tcgetattr(0, old_term);
+	if (ret)
+		return EFI_LOAD_ERROR;
+
+	raw_term = *old_term;
+	cfmakeraw(&raw_term);
+	raw_term.c_oflag |= (ONLCR | OPOST);
```
Comment (jeremy-compostella) [inline host/terminal_conin.c]: unnecessary parenthesis.

File: host/terminal_conin.c [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+static struct termios *old_term;
+
+static EFI_STATUS makeraw_term(VOID)
+{
+	int ret;
+	struct termios raw_term;
+
+	old_term = malloc(sizeof(*old_term));
+	if (old_term == NULL)
```
Comment (jeremy-compostella) [inline host/terminal_conin.c]: coding style consistency
if (!old_term)

File: host/terminal_conin.c [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+static EFIAPI VOID
+terminal_conin_wait_for_key(__attribute__((__unused__)) EFI_EVENT Event,
+			    VOID *Context)
+{
+	int n;
+	char buf[3];
+	EFI_SYSTEM_TABLE *st = (EFI_SYSTEM_TABLE *)Context;
+
+	n = read(STDIN_FILENO, buf, sizeof(buf)*sizeof(buf[0]));
```
Comment (jeremy-compostella) [inline host/terminal_conin.c]: Why do you read 3 characters at a time ?

File: host/event.c [hunk 8]
```diff
@@ -118,6 +128,14 @@ wait_for_event(UINTN NumberOfEvents,
 	if (ret)
 		return EFI_DEVICE_ERROR;
 
+	if (event->type == EVT_NOTIFY_WAIT) {
+		ret = pthread_create(&thread_notify, NULL, call_notify, event);
+		if (ret) {
+			pthread_mutex_unlock(&event->lock);
+			return EFI_DEVICE_ERROR;
```
Comment (jeremy-compostella) [inline host/event.c:135]: Where do you call pthread_join to clear the thread data ?

**Review by jeremy-compostella — COMMENTED** [review]
It sounds very promising.  I only have a couple of minor comment.

Hope you are doing well.
Thanks, I appreciate you did that.

File: host/image.h (hunk 1 above)
Comment (ghost) [reply host/image.h:41]: Actually to make the change smaller, setjmp() and longjmp() function wrapper from gnuefi lib need a pointer to jmp_buf as parameter.
If you really dont like it, I can do the change on the callers.

File: host/main.c (hunk 2 above)
Comment (ghost) [reply host/main.c:74]: Same reason here.

File: host/terminal_conin.c (hunk 7 above)
Comment (ghost) [reply host/terminal_conin.c]: It is to properly consume ANSI escape sequences \033[A ... \033[D emited when you press arrow keys.
Maybe I can make a comment here or a #define ANSI_ESCAPE_LEN 3, which one do you prefer?

File: host/event.c (hunk 8 above)
Comment (ghost) [reply host/event.c:135]: Good catch I forgot to detach the thread.

File: host/terminal_conin.c (hunk 7 above)
Comment (jeremy-compostella) [reply host/terminal_conin.c]: OK.  Make sure it is understandable/obvious for whoever will read the code.

**Comment by ghost** [comment]
Hi,

I've updated new patches according to your remarks.
I've detached the thread_notify on event.c and fixed a memory leak in terminal_conin.c (forgot to call CloseEvent on the exit function driver).
Please note that I've added an extra commit in the pull request (sha: 7d00097) please review it also.
Please let me know about the "jmp_buf jmp[1]" open.
I have to check that I did not broke the Android build before doing a pull request on 01.org, by the way are you still in charge of the 01.org upstream repo?

File: host/terminal_conin.c (hunk 7 above)
Comment (jeremy-compostella) [reply host/terminal_conin.c]: One more thing, coding style:
sizeof(buf)*sizeof(buf[0]));  -> sizeof(buf) * sizeof(buf[0]));

**Review by jeremy-compostella — COMMENTED** [review]
Minor comments.

I am not in charge of the 01org upstream even though I still have access.  If you want to get your change pushed to the Intel repository, you should contact "Chen, Xihua" <xihua.chen@intel.com> and CC me.

Also, let me know when you are done with the "Android" tests.

**Pull request merged**